### ISSUE TITLE
Can't parse snowflake create view DDL

### ISSUE THREAD

**Issue by zhaidewei** [sqlfluff/sqlfluff#5016]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

Below Snowflake SQL is valid:
```sql
create or replace view SCHEMANAME.TABLENAME(
COL1,
COL2
) as
SELECT
$1 COL1, 
$2 COL2
FROM @NAME/path/(file_format=>CUSTOMIZED_FILE_FORMAT,pattern=>'.*ZIG37426_.*labels_.*.txt.gz.*')
;
```
But `sqlfluff lint` showed unparsable section

### Expected Behaviour

It can be parsed.

### Observed Behaviour

```bash
sqlfluff lint output/sample.sql --config ./.sqlfluff
== [output/sample.sql] FAIL
L:   8 | P:  96 |  PRS | Line 8, Position 96: Found unparsable section: ')\n;'
WARNING: Parsing errors found and dialect is set to 'snowflake'. Have you configured your dialect correctly?
All Finished 📜 🎉!
```

### How to reproduce

1. save the sample sql into file `example.sql`
2. use the sample `.sqlfluff` file
3. Run `sqlfluff lint` against `example.sql`

### Dialect

snowflake

### Version

2.1.4

### Configuration

[sqlfluff]
dialect = snowflake
large_file_skip_byte_limit = 0
exclude_rules = all

[sqlfluff:templater:jinja:context]
database_name = dummy
account = dummy
environment = prd

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by jamiekt** [comment]
I'm not a sqlfluff maintainer but I'd quite like to understand the sqlfluff codebase a bit better so I started poking around to try and figure out this issue. I did not (yet) figure out the root cause, let alone a fix, but I did discover that the error goes away when the @ at the start of the FROM clause is removed. Of course, that doesn't help because its no longer a valid statement, the "@" is required for this view.

I changed it to 

```sql
create or replace view SCHEMANAME.TABLENAME(
COL1,
COL2
) as
SELECT
$1 COL1, 
$2 COL2
FROM `@NAME/path/(file_format=>CUSTOMIZED_FILE_FORMAT,pattern=>'.*ZIG37426_.*labels_.*.txt.gz.*')`
;
```
and the error went away however I don't know if that is valid syntax or not, perhaps the OP can let us know?

**Comment by andychannery** [comment]
I was looking at [Snowflake documentation](https://docs.snowflake.com/en/user-guide/querying-stage#query-examples) regarding querying data from staged files and it seems that the query syntax requires a space between the storage location and the file format/pattern statement (refer to query examples in documentation).

That is:
```
create or replace view SCHEMANAME.TABLENAME(
COL1,
COL2
) as
SELECT
$1 COL1, 
$2 COL2
FROM @NAME/path/ (file_format=>CUSTOMIZED_FILE_FORMAT,pattern=>'.*ZIG37426_.*labels_.*.txt.gz.*')
;
```

With the addition of the space this becomes error free. Was this the issue @zhaidewei ?

**Comment by moreaupascal56** [comment]
Hello there, I think @andychannery is right the bettr way ould be to have a space between the stage_path and the file format specifications. 

I was curious about why it fails sqlfluff wise so I had a look and wrote this explanation for those that may be interested :). (I tried to be clear but it is a bit messy) 

A good way to start is to take the script that is working with the space (from @andychannery) and try to understand what is going on, for that add the working query on the [create_view.sql file](https://github.com/sqlfluff/sqlfluff/blob/main/test/fixtures/dialects/snowflake/create_view.sql) then generate the yml test files with ` tox -e generate-fixture-yml  `, it will update  [create_view.yml](https://github.com/sqlfluff/sqlfluff/blob/main/test/fixtures/dialects/snowflake/create_view.yml) and results in this: 
```sql
create or replace view SCHEMANAME.TABLENAME(
COL1,
COL2
) as
SELECT
$1 COL1,
$2 COL2
FROM @NAME/path/(file_format=>CUSTOMIZED_FILE_FORMAT,pattern=>'.*ZIG37426_.*labels_.*.txt.gz.*')
;

```
this sql query becomes:

```yml

- statement:
    create_view_statement:             # The parser recognizes a create view statement
    - keyword: create
    - keyword: or
    - keyword: replace
    - keyword: view
    - table_reference:
      - naked_identifier: SCHEMANAME
      - dot: .
      - naked_identifier: TABLENAME
    - bracketed:
      - start_bracket: (
      - column_reference:
          naked_identifier: COL1
      - comma: ','
      - column_reference:
          naked_identifier: COL2
      - end_bracket: )
    - keyword: as
    - select_statement:
        select_clause:
        - keyword: SELECT
        - select_clause_element:
            column_reference:
              column_index_identifier_segment: $1
            alias_expression:
              naked_identifier: COL1
        - comma: ','
        - select_clause_element:
            column_reference:
              column_index_identifier_segment: $2
            alias_expression:
              naked_identifier: COL2
        from_clause:                              # Here is the part we are interested in 
          keyword: FROM
          from_expression:
            from_expression_element:
              table_expression:
                stage_path: '@NAME/path/'
                bracketed:
                - start_bracket: (
                - keyword: file_format
                - parameter_assigner: =>
                - file_format_segment:
                    object_reference:
                      naked_identifier: CUSTOMIZED_FILE_FORMAT
                - comma: ','
                - keyword: pattern
                - parameter_assigner: =>
                - quoted_literal: "'.*ZIG37426_.*labels_.*.txt.gz.*'"
                - end_bracket: )
- statement_terminator: ;

```

The yaml file is the output of the sql parser from SQLfluff, each part of the query is matched with a Statement correspondign to the object. 

You can see that 
Let's have a closer look at the part we care about: 

```yml
        from_clause:                                                                                     
          keyword: FROM
          from_expression:
            from_expression_element:
              table_expression:
                stage_path: '@NAME/path/'
                bracketed:
                - start_bracket: (
                - keyword: file_format
                - parameter_assigner: =>
                - file_format_segment:
                    object_reference:
                      naked_identifier: CUSTOMIZED_FILE_FORMAT
                - comma: ','
                - keyword: pattern
                - parameter_assigner: =>
                - quoted_literal: "'.*ZIG37426_.*labels_.*.txt.gz.*'"
                - end_bracket: )


```


the keys of the yaml files are the statement type like here for the `FileFormatSegment` it is `file_format_segment`
![Capture d’écran 2023-10-09 à 15 49 44](https://github.com/sqlfluff/sqlfluff/assets/50910271/e1640044-723f-4c2a-a2dd-3c0ad42fa67b). 

In the table_expression, we can see that it recognizes a stage_path type of statement and the file format segment. 
Let's have a look: 
-  table_expression: is a global statement from [dialect_ansi.py
](https://github.com/sqlfluff/sqlfluff/blob/4d5af85abe6a8da0b4c8aae3d22e799af08d37d9/src/sqlfluff/dialects/dialect_ansi.py#L1528C2-L1528C2) BUT it is overwrite in [dialect_snowflake.py
](https://github.com/sqlfluff/sqlfluff/blob/4d5af85abe6a8da0b4c8aae3d22e799af08d37d9/src/sqlfluff/dialects/dialect_snowflake.py#L4988) so we have to consider this one, I copied it below: 

```python
class TableExpressionSegment(ansi.TableExpressionSegment):
    """The main table expression e.g. within a FROM clause."""

    match_grammar = OneOf(
        Ref("BareFunctionSegment"),
        Ref("FunctionSegment"),
        Ref("TableReferenceSegment"),
        # Nested Selects
        Bracketed(Ref("SelectableGrammar")),
        Ref("ValuesClauseSegment"),
        Sequence(                                         # HERE :D 
            Ref("StagePath"),
            Bracketed(
                Delimited(
                    Sequence(
                        "FILE_FORMAT",
                        Ref("ParameterAssignerSegment"),
                        Ref("FileFormatSegment"),
                    ),
                    Sequence(
                        "PATTERN",
                        Ref("ParameterAssignerSegment"),
                        Ref("QuotedLiteralSegment"),
                    ),
                ),
                optional=True,
            ),
        ),
    )

```

Here we see that the Sequence object is mentioning a StagePath Element which is defined [here](https://github.com/sqlfluff/sqlfluff/blob/4d5af85abe6a8da0b4c8aae3d22e799af08d37d9/src/sqlfluff/dialects/dialect_snowflake.py#L332): 
```python
    StagePath=RegexParser(
        r"(?:@[^\s;)]+|'@[^']+')",
        ansi.IdentifierSegment,
        type="stage_path",
    ),

```

As there is no spaces in the example provided below it considers that all the line: `@NAME/path/(file_format=>CUSTOMIZED_FILE_FORMAT,pattern=>'.*ZIG37426_.*labels_.*.txt.gz.*')`should be recognize in this regex. Indeed a Sequence means a sequence of statement separated by a space basically, so this is why in the example from @andychannery it goes in the FileFormat part of the TableExpressionSegment. 

And when testing the REGEX
![Capture d’écran 2023-10-09 à 16 00 09](https://github.com/sqlfluff/sqlfluff/assets/50910271/b4a3f6c5-1478-4522-82ee-910c654c5714)
You can see that the ending parenthesis is not recognize, this is why we have the error message (saying that the parenthesis is the issue) below with SQLFLUFF
 
<img width="779" alt="Capture d’écran 2023-10-09 à 16 00 45" src="https://github.com/sqlfluff/sqlfluff/assets/50910271/88faa74a-90ed-4edf-980d-adc08fa2f01d">


SO basically I think the easiest way for you is to add this space, but if Snowflake works with your initial syntax we should maybe update the StagePath SEgment. However I do not see how we could update it to accept your syntax while still checking the file format element (because there is no clear separation between the stage_path and the file_format. 

PS: Another way to dig these issues is to start from the [CreateVIewStatement](https://github.com/sqlfluff/sqlfluff/blob/4d5af85abe6a8da0b4c8aae3d22e799af08d37d9/src/sqlfluff/dialects/dialect_snowflake.py#L4135) and go from there. I think this way should be used for simpler queries, here as it was pretty layered and deep it was easier to go with the YML file :)

**Comment by peterbud** [comment]
Closing, as the syntax _requires_ a space between the storage location and the file format/pattern
